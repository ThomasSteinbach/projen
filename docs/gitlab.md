# GitLab

[GitLab CI/CD](https://docs.gitlab.com/ee/ci/yaml/gitlab_ci_yaml.html) uses a `.gitlab-ci.yml` file located at the root of your repository to specify continuous integration and continuous deployment processes. GitLab CI configuration files can be split into multiple files using the [include](https://docs.gitlab.com/ee/ci/yaml/#include) fields in order to increase readability or reduce duplication of the same configuration in multiple places.

## Creating a configuration

We support the creation of GitLab CI/CD Configurations through a `CiConfiguration` class and a higher level `GitlabConfiguration` and `NestedConfiguration` classes. To create a GitLab configuration you must first create a [project](./project.md) to assign the configuration to.

When using GitLab for CI, the `.gitlab-ci.yml` is always located at the root of the project while other configuration files are located in the `.gitlab\ci-templates\` subdirectory.

```js
const { typescript, gitlab } = require('projen');
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: 'main',
  name: 'my-project',
});
const gitlabMain = new gitlab.GitlabConfiguration(project,
  {
    jobs: {
      myJob: {
        script: 'echo Hello',
      },
    },
  });
gitlabMain.createNestedTemplates({
  myNestedTemplate: {
    jobs: {
      myOtherJob: {
        script: 'echo World!',
      },
    },
  },
});
project.synth();
```

This creates the following directory structure.

```shell
├── .gitlab-ci.yml
└── .gitlab
    └── ci-templates
         └── mynestedtemplate.yml
```

## Updating Nested Configurations

You can modify a `NestedConfiguration` through the `nestedTemplates` property of a `GitlabConfiguration`.

```js
const { typescript, gitlab } = require('projen');
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: 'main',
  name: 'my-project',
});
const gitlabMain = new gitlab.GitlabConfiguration(project,
  {
    jobs: {
      myJob: {
        script: 'echo Hello',
      },
    },
  });
gitlabMain.createNestedTemplates({
  myNestedTemplate: {},
});
gitlabMain.nestedTemplates.myNestedTemplate.addStages('stage');
project.synth();
```

## Sequences

Sequences are not part of the standard Gitlab CI definition, but are a powerful addition of the projen Gitlab library.
They allow you to bundle multiple jobs for:

* applying common configuration to them.
* re-using the jobs within a pipeline, e.g. running the same kind of jobs for multiple stages.
* providing pipeline snippets, consisting of multiple jobs doing a specific job, in your own library.

Let's show a short example, how how Sequences work:

```typescript
import { GitlabConfiguration, Image, Sequence } from "../../src/gitlab"
import { synthSnapshot, TestProject } from '../util'

test("create documentation example project", () => {


    class PublishApp extends Sequence {
        public constructor(environment: string) {
            super()
            this.stageToPrepend = environment
            this.addChildren({
                job1: { script: ["build my app"], stage: "build" },
                job2: { script: [`upload to ${environment}`], stage: "deploy" }
            })
        }
    }

    const project = new TestProject()
    const config = new GitlabConfiguration(project)

    const publish_dev = new PublishApp("dev")
    const publish_prd = new PublishApp("prd")

    const buildImage: Image = { name: "mycompany/buildimage:latest" }
    publish_dev.imageToInit = buildImage
    publish_prd.imageToInit = buildImage

    publish_dev.tagsToAdd = ["dev-runner"]
    publish_prd.tagsToAdd = ["prd-runner"]

    config.addJobs({ dev: publish_dev, prd: publish_prd })

    expect(synthSnapshot(project)[".gitlab-ci.yml"]).toMatchSnapshot();
})
```

We have wrapped the code of the `projen.ts` into a Jest test file. This allows us and you to simply generate a snapshot file with the contents of the `.gitlab-ci.yaml` by running:

```sh
npx jest test/gitlab/docs-sequence.test.ts -t 'create documentation example project' -u
```

The resulting pipeline should look like following:

```yaml
stages:
  - devbuild
  - devdeploy
  - prdbuild
  - prddeploy
dev_job1:
  image:
    name: mycompany/buildimage:latest
  script:
    - build my app
  stage: devbuild
  tags:
    - dev-runner
dev_job2:
  image:
    name: mycompany/buildimage:latest
  script:
    - upload to dev
  stage: devdeploy
  tags:
    - dev-runner
prd_job1:
  image:
    name: mycompany/buildimage:latest
  script:
    - build my app
  stage: prdbuild
  tags:
    - prd-runner
prd_job2:
  image:
    name: mycompany/buildimage:latest
  script:
    - upload to prd
  stage: prddeploy
  tags:
    - prd-runner
```

Let's sum up some aspects of sequences we have used in this example:

* We have provided a reusable sequence of jobs by providing the `PublishApp` class.
  The jobs within that sequence do a common task: building and publishing our application.
  Those jobs are re-usable because we spent our sequence class a configurable interface by the constructor arguments.
  We also could have created a function returning a sequence, but a class is mostly a better choice as it provides
  more possibilities, like storing additional information or publishing parts of the contents.
* We used the common configuration interface of the sequence to modify the contained jobs.
  We do this without taking care of which jobs are inside the sequence.
  So we conveniently configured the the stage, build image and tags.
* The sequence takes care of preventing naming conflicts between jobs, by namespacing the jobs depending on the sequence they are in.

The configuration interface of sequences allows to modify all properties of the contained jobs.
For this the sequence provide methods that starts with the name of the job property and end with following names:

* `*ToInit` - Set a value for that property on every job whose value hasn't been set before.
* `*ToReplace` - Replaces the property value on every job.
* `*ToAdd` - Adds a value to a job property where possible, like on `Job.artifacts`.
* `*ToPrepend` - Inserts a value before any existing values of that job property where possible, like on `Job.script`.
* `*ToAppend` - Inserts a value after any existing values of that job property where possible, like on `Job.script`.

For more information on the configuration interface please consult the Sequences constructor API interface.

The good things don't end here: Sequences could not only contain jobs but sequences itself.
This allows you to compose your pipeline of multiple job sequences vertically and horizontally.
Modifications are applied recursively to all jobs in all sequences.
Especially this allows you to use predefined sequences of others and apply your configurations as needed.
