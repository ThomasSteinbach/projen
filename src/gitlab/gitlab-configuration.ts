import { CiConfigurationOptions } from ".";
import { Project } from "../project";
import { CiConfiguration } from "./configuration";
import { NestedConfiguration } from "./nested-configuration";

/**
 * A GitLab CI for the main `.gitlab-ci.yml` file.
 */
export class GitlabConfiguration extends CiConfiguration {
  public readonly nestedTemplates: Record<string, NestedConfiguration> = {};
  constructor(project: Project, options?: CiConfigurationOptions) {
    super(project, "gitlab-ci", options);
  }

  /**
   * Before doing the rendering of the super class, we update all stages
   * from nested tamplates, that were added after the nested template
   * has been added to this CI configuration.
   */
  protected renderCI() {
    Object.values(this.nestedTemplates).forEach(template => {
      this.addStages(...template.stages);
    })
    return super.renderCI()
  }

  /**
   * Creates and adds nested templates to the includes of the main CI.
   * Additionally adds their stages to the main CI if they are not already present.
   * You can futher customize nested templates through the `nestedTemplates` property.
   * E.g. gitlabConfig.nestedTemplates['templateName']?.addStages('stageName')
   * @param config a record the names and configuraitons of the templates.
   */
  public createNestedTemplates(config: Record<string, CiConfigurationOptions>) {
    for (const [name, options] of Object.entries(config)) {
      if (this.nestedTemplates[name] !== undefined) {
        throw new Error(
          `${this.name}: GitLab CI already contains template "${name}".`
        );
      }
      const template = new NestedConfiguration(
        this.project,
        this,
        name,
        options
      );
      this.nestedTemplates[template.name] = template;
      this.addIncludes({ local: template.path });
      this.addStages(...template.stages);
    }
  }
}
