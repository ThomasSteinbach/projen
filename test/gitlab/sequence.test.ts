import { GitlabConfiguration } from '../../src/gitlab'
import { Artifacts, Cache, CachePolicy, CacheWhen, DefaultElement, Job, KubernetesEnum } from '../../src/gitlab/configuration-model'
import { Sequence } from '../../src/gitlab/sequence'
import { synthSnapshot, TestProject } from '../util'

/**
 * We use `toStrictEqual` instead of `toEqual` because we don't want that
 * a Sequence will introduce new `Job` keys with `undefined` values. So
 * `toStrictEqual` will also compare those keys.
 */

test("Test adding an element to Sequence with existing name", () => {
    const sequence = new Sequence()
    sequence.addChildren({ foo: {} })
    expect(() => sequence.addChildren({ foo: {} }))
        .toThrowError("A child was added to a Sequence with a name that already exsits")
})

/**
 * All `Job` values whose type is a list of values are modified the
 * same way within a `Sequence` by the private function `modify_list`.
 * 
 * This test proves this function works on an empty Sequence, with
 * no jobs.
 * 
 * We only test this function for the job value `afterScript`, because
 * it should behave the same for:
 * 
 * * afterScript
 * * beforeScript
 * * dependencies
 * * extends
 * * needs
 * * rules
 * * script
 * * services
 * * tags
 */
test("Test private function 'modify_list' on empty Sequence", () => {
    const sequence = new Sequence()
    expect(sequence.render()).toStrictEqual({})

    const testArray = ["a"]

    sequence.afterScriptToReplace = testArray
    expect(sequence.render()).toStrictEqual({})

    sequence.afterScriptToPrepend = testArray
    expect(sequence.render()).toStrictEqual({})

    sequence.afterScriptToAppend = testArray
    expect(sequence.render()).toStrictEqual({})
})

/**
 * All `Job` values whose type is a list of values are modified the
 * same way within a `Sequence` by the private function `modify_list`.
 * 
 * This test proves this function works on a Sequence, containing jobs.
 * 
 * We only test this function for the job value `afterScript` and
 * `dependencies` because it should behave the same for:
 * 
 * * afterScript
 * * beforeScript
 * * dependencies
 * * extends
 * * needs
 * * rules
 * * script
 * * services
 * * tags
 * 
 * The difference between `afterScript` and `dependencies` is that the
 * `modify_list` has an overloaded function signature. Both cases are
 * tested.
 */
test("Test private function 'modify_list'", () => {

    const sequence = new Sequence()
    sequence.addChildren({ job1: {} })
    sequence.addChildren({ job2: { afterScript: ["x"], dependencies: ["y"] } })
    expect(sequence.render()).toStrictEqual({ job1: {}, job2: { afterScript: ["x"], dependencies: ["y"] } })

    sequence.afterScriptToPrepend = ["a"]
    expect(sequence.render()).toStrictEqual(
        {
            job1: { afterScript: ["a"] },
            job2: { afterScript: ["a", "x"], dependencies: ["y"] }
        }
    )

    sequence.afterScriptToAppend = ["b"]
    sequence.dependenciesToAdd = ["1"]
    expect(sequence.render()).toStrictEqual(
        {
            job1: { afterScript: ["a", "b"], dependencies: ["1"] },
            job2: { afterScript: ["a", "x", "b"], dependencies: ["1", "y"] }
        }
    )

    sequence.afterScriptToReplace = ["c"]
    sequence.dependenciesToReplace = ["2"]
    expect(sequence.render()).toStrictEqual(
        {
            job1: { afterScript: ["c"], dependencies: ["2"] },
            job2: { afterScript: ["c"], dependencies: ["2"] }
        }
    )
})

test("Test 'allowFailure' modifications", () => {
    const sequence = new Sequence()

    // test both cases, where the original allowFailure is a boolean
    // or an AllowFailure object
    sequence.addChildren({ job1: { allowFailure: true } })
    sequence.addChildren({ job2: { allowFailure: { exitCodes: 1 } } })

    expect(sequence.render()).toStrictEqual(
        {
            job1: { allowFailure: true },
            job2: { allowFailure: { exitCodes: 1 } }
        }
    )

    sequence.allowFailureToAdd = { exitCodes: 3 }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { allowFailure: { exitCodes: 3 } },
            job2: { allowFailure: { exitCodes: [1, 3] } }
        }
    )

    sequence.allowFailureToReplace = false
    expect(sequence.render()).toStrictEqual(
        {
            job1: { allowFailure: false },
            job2: { allowFailure: false }
        }
    )

    sequence.allowFailureToReplace = { exitCodes: 4 }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { allowFailure: { exitCodes: 4 } },
            job2: { allowFailure: { exitCodes: 4 } }
        }
    )
})

test("Test 'artifacts' modifications", () => {
    const sequence = new Sequence()

    const originalArtifacts: Artifacts = {
        exclude: ["original"],
        expireIn: "original",
        exposeAs: "original",
        name: "original",
        paths: ["original"],
        reports: { performance: ["original"] },
        untracked: true,
        when: CacheWhen.ALWAYS,
    }

    sequence.addChildren({ job1: { artifacts: originalArtifacts } })
    expect(sequence.render()).toStrictEqual({ job1: { artifacts: originalArtifacts } })

    sequence.artifactsToAdd = {
        exclude: ["modified"],
        expireIn: "modified",
        exposeAs: "modified",
        name: "modified",
        paths: ["modified"],
        reports: { junit: ["modified"] },
        untracked: false,
        when: CacheWhen.ON_FAILURE,
    }
    expect(sequence.render()).toStrictEqual({
        job1: {
            artifacts: {
                exclude: ["original", "modified"],
                expireIn: "modified",
                exposeAs: "modified",
                name: "modified",
                paths: ["original", "modified"],
                reports: { junit: ["modified"] },
                untracked: false,
                when: CacheWhen.ON_FAILURE,
            }
        }
    })
})

test("Test 'cache' modifications", () => {
    const sequence = new Sequence()

    const originalCache: Cache = {
        paths: ["original"],
        key: "original",
        untracked: true,
        policy: CachePolicy.PULL,
        when: CacheWhen.ALWAYS,
    }

    sequence.addChildren({ job1: { cache: originalCache } })
    expect(sequence.render()).toStrictEqual({ job1: { cache: originalCache } })

    sequence.cacheToAdd = {
        paths: ["modified"],
        key: "modified",
        untracked: false,
        policy: CachePolicy.PULL_PUSH,
        when: CacheWhen.ON_SUCCESS,
    }
    expect(sequence.render()).toStrictEqual({
        job1: {
            cache: {
                paths: ["original", "modified"],
                key: "modified",
                untracked: false,
                policy: CachePolicy.PULL_PUSH,
                when: CacheWhen.ON_SUCCESS,
            }
        }
    })
})

test("Test 'cache key' modifications", () => {
    const sequence = new Sequence()

    const originalCache: Cache = {
        key: { files: ["original"], prefix: "foo" }
    }
    sequence.addChildren({ job1: { cache: originalCache } })
    expect(sequence.render()).toStrictEqual({ job1: { cache: originalCache } })

    sequence.cacheToAdd = { key: { files: ["modified"], prefix: "bar" } }
    expect(sequence.render()).toStrictEqual({
        job1: {
            cache: {
                key: {
                    files: ["original", "modified"],
                    prefix: "bar"
                }
            }
        }
    })
})

test("Test 'except' modifications", () => {

    /**
     * Those tests should also be valid for 'only' modifications.
     */

    const sequence = new Sequence()

    sequence.addChildren({ job1: { except: ["original"] } })
    sequence.addChildren({
        job2: {
            except: {
                changes: ["original"],
                refs: ["original"],
                variables: ["original"],
            }
        }
    })
    expect(sequence.render()).toStrictEqual(
        {
            job1: { except: ["original"] },

            job2: {
                except: {
                    changes: ["original"],
                    refs: ["original"],
                    variables: ["original"],
                }
            }
        }
    )

    sequence.exceptToAdd = ["modified"]
    expect(sequence.render()).toStrictEqual(
        {
            job1: { except: ["original", "modified"] },

            job2: {
                except: {
                    changes: ["original"],
                    refs: ["original", "modified"],
                    variables: ["original"],
                }
            }
        }
    )

    sequence.exceptToAdd = {
        changes: ["modified"],
        kubernetes: KubernetesEnum.ACTIVE,
        refs: ["modified"],
        variables: ["modified"],
    }
    expect(sequence.render()).toStrictEqual(
        {
            job1: {
                except: {
                    changes: ["modified"],
                    kubernetes: KubernetesEnum.ACTIVE,
                    refs: ["original", "modified"],
                    variables: ["modified"],
                }
            }
            ,

            job2: {
                except: {
                    changes: ["original", "modified"],
                    kubernetes: KubernetesEnum.ACTIVE,
                    refs: ["original", "modified"],
                    variables: ["original", "modified"],
                }
            }
        }
    )
})

test("Test 'inherit' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: { inherit: { default: false, variables: false } } })
    sequence.addChildren({ job2: { inherit: { default: [DefaultElement.IMAGE], variables: ["original"] } } })
    expect(sequence.render()).toStrictEqual(
        {
            job1: { inherit: { default: false, variables: false } },
            job2: { inherit: { default: [DefaultElement.IMAGE], variables: ["original"] } }
        },
    )

    sequence.inheritToAdd = { default: true, variables: true }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { inherit: { default: true, variables: true } },
            job2: { inherit: { default: true, variables: true } }
        },
    )

    sequence.inheritToAdd = { default: [DefaultElement.CACHE], variables: ["modified"] }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { inherit: { default: [DefaultElement.CACHE], variables: ["modified"] } },
            job2: { inherit: { default: [DefaultElement.IMAGE, DefaultElement.CACHE], variables: ["original", "modified"] } }
        },
    )
})

test("Test 'parallel' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: { parallel: 1 } })
    sequence.addChildren({ job2: { parallel: { matrix: [{ foo: ["bar"] }] } } })
    expect(sequence.render()).toStrictEqual(
        {
            job1: { parallel: 1 },
            job2: { parallel: { matrix: [{ foo: ["bar"] }] } }
        }
    )

    const originalConsoleWarn = global.console.warn
    global.console.warn = jest.fn()

    sequence.parallelToAdd = { matrix: [{ hello: ["world"] }] }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { parallel: 1 },
            job2: { parallel: { matrix: [{ foo: ["bar"] }, { hello: ["world"] }] } }
        }
    )

    expect(console.warn).toBeCalledWith("Incompatible types for 'Job.parallel' and 'Sequence.parallelToAdd'. Ignoring Job and keeping its original 'parallel' value.")
    expect(console.warn).toBeCalledTimes(1)
    global.console.warn = originalConsoleWarn
})

test("Test 'secrets' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: { secrets: { secret1: { key1: { vault: { engine: { name: "a", path: "b" }, field: "c", path: "d" } } } } } })
    sequence.addChildren({ job2: { secrets: { secret1: { key2: { vault: { engine: { name: "e", path: "f" }, field: "g", path: "h" } } } } } })
    sequence.addChildren({ job3: { secrets: { secret2: { key1: { vault: { engine: { name: "i", path: "j" }, field: "k", path: "l" } } } } } })
    expect(sequence.render()).toStrictEqual(
        {
            job1: { secrets: { secret1: { key1: { vault: { engine: { name: "a", path: "b" }, field: "c", path: "d" } } } } },
            job2: { secrets: { secret1: { key2: { vault: { engine: { name: "e", path: "f" }, field: "g", path: "h" } } } } },
            job3: { secrets: { secret2: { key1: { vault: { engine: { name: "i", path: "j" }, field: "k", path: "l" } } } } }
        },
    )

    sequence.secretsToAdd = { secret1: { key1: { vault: { engine: { name: "w", path: "x" }, field: "y", path: "z" } } } }
    expect(sequence.render()).toStrictEqual(
        {
            job1: { secrets: { secret1: { key1: { vault: { engine: { name: "w", path: "x" }, field: "y", path: "z" } } } } },

            job2: {
                secrets: {
                    secret1: {
                        key2: { vault: { engine: { name: "e", path: "f" }, field: "g", path: "h" } },
                        key1: { vault: { engine: { name: "w", path: "x" }, field: "y", path: "z" } }
                    }
                }
            },
            job3: {
                secrets: {
                    secret2: { key1: { vault: { engine: { name: "i", path: "j" }, field: "k", path: "l" } } },
                    secret1: { key1: { vault: { engine: { name: "w", path: "x" }, field: "y", path: "z" } } }
                }
            }
        },
    )
})

test("Test 'stage' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: {} })
    sequence.addChildren({ job2: { stage: "original" } })
    expect(sequence.render()).toStrictEqual({ job1: {}, job2: { stage: "original" } })

    sequence.stageToPrepend = "modified"
    expect(sequence.render()).toStrictEqual({ job1: { stage: "modified" }, job2: { stage: "modifiedoriginal" } })
})

test("Test 'variables' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: {} })
    sequence.addChildren({ job2: { variables: { foo: "original", bar: 1 } } })
    expect(sequence.render()).toStrictEqual({ job1: {}, job2: { variables: { foo: "original", bar: 1 } } })
})

test("Test '*ToInit' modifications", () => {
    const sequence = new Sequence()

    sequence.addChildren({ job1: {}, job2: { afterScript: ["b"], allowFailure: false, interruptible: false } })
    expect(sequence.render()).toStrictEqual({ job1: {}, job2: { afterScript: ["b"], allowFailure: false, interruptible: false } })

    /**
     * The `ToInit*` modifications works for all variables the same but for variables, that
     * hold boolean variables. Latter are 'allowFailure' and 'interruptible'. Thus we test
     * those two values explicitly.
     */
    sequence.afterScriptToInit = ["a"]
    sequence.allowFailureToInit = true
    sequence.interruptibleToInit = true
    expect(sequence.render()).toStrictEqual(
        {
            job1: { afterScript: ["a"], allowFailure: true, interruptible: true },
            job2: { afterScript: ["b"], allowFailure: false, interruptible: false }
        })
})

test("Test job names of sub sequences are prefixed", () => {
    const sequence = new Sequence()
    const subsequence = new Sequence()

    sequence.addChildren({ a: {}, b: subsequence })
    subsequence.addChildren({ c: {} })

    expect(sequence.render()).toStrictEqual({ a: {}, b_c: {} })
})

test("Test stage ordering from jobs sequences and nested templates", () => {
    const project = new TestProject({ stale: true })
    const ciconfig = new GitlabConfiguration(project)

    /**
     * Stages from jobs that are already present, are added to the CI config in that
     * moment the job is added to the CI config. This is valid for plain jobs as well
     * as for jobs from sequences and nested templates.
     */

    ciconfig.addStages("stage1")

    ciconfig.createNestedTemplates({ foo: { jobs: { bar: { stage: "stage2" } } } });

    const sequence1 = new Sequence()
    sequence1.addChildren({ job1: { stage: "stage3" } })

    /**
     * For subsequences it shouldn't make a difference, when the subsequence
     * is added to the sequence, as long as the sequence isn't added to the
     * CI config before. Stages within sequences just evaluated when been added
     * to the CI config and then read vertically first from the sequence tree.
     * We prove that case with sequence2 and sequence2_2.
     */
    const sequence1_1 = new Sequence()
    sequence1_1.addChildren({ job1: { stage: "stage4" } })
    sequence1.addChildren({ seq1: sequence1_1 })

    // 'stage5' is defined in a plain job below

    const sequence2 = new Sequence()
    sequence2.addChildren({ job1: { stage: "stage7" } })

    const sequence2_1 = new Sequence()
    sequence2_1.addChildren({ job1: { stage: "stage8" } })
    sequence2.addChildren({ seq1: sequence2_1 })

    /**
     * This job has stage #5 even when he is added to sequence2_1 first.
     * As said above, stages from sequences are evaluated when been added
     * to the CI configuration and then vertically first. Thus this job
     * comes right after the job from sequence1_1.
     */
    const jobx: Job = { stage: "stage5" }
    sequence2_1.addChildren({ jobx: jobx })
    sequence1.addChildren({ jobx: jobx })

    /**
     * Now we add all jobs, sequences and templates (that haven't already been added)
     * to the CI config.
     */
    ciconfig.addJobs({
        bar: sequence1,
        job1: { stage: "stage6" },
        baz: sequence2,
    })
    ciconfig.createNestedTemplates({ baz: { jobs: { bar: { stage: "stage9" } } } });

    /**
     * However stages and jobs added to nested templates and sequences after those
     * templates and sequences have been added to the CI config are also added to
     * the CI config afterwards. First the stages of the nested temples were added,
     * second the stages of the sequences. Within templates and sequences the stages
     * were added in the same order, the templates/sequences were added to the CI
     * configuration.
     */

    ciconfig.nestedTemplates["baz"].addStages("stage11")
    sequence2.addChildren({ job2: { stage: "stage14" } })
    ciconfig.nestedTemplates["foo"].addStages("stage10")
    sequence1.addChildren({ job2: { stage: "stage12" } })

    /**
     * Same as with 'jobx' but now after previous stages have been added
     * to the ci config.
     */
    const joby: Job = { stage: "stage13" }
    sequence2_1.addChildren({ joby: joby })
    sequence1.addChildren({ joby: joby })

    /**
     * This test also prooves that all jobs were correctly added
     * to the final CI configuration. Thus we check it matches the snapshot.
     */
    expect(synthSnapshot(project)[".gitlab-ci.yml"]).toMatchSnapshot();
    expect(ciconfig.stages).toStrictEqual([
        "stage1",
        "stage2",
        "stage3",
        "stage4",
        "stage5",
        "stage6",
        "stage7",
        "stage8",
        "stage9",
        "stage10",
        "stage11",
        "stage12",
        "stage13",
        "stage14",
    ])
})