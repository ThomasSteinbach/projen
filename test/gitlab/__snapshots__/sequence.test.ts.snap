// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test stage ordering from jobs sequences and nested templates 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

include:
  - local: .gitlab/ci-templates/foo.yml
  - local: .gitlab/ci-templates/baz.yml
stages:
  - stage1
  - stage2
  - stage3
  - stage4
  - stage5
  - stage6
  - stage7
  - stage8
  - stage9
  - stage10
  - stage11
  - stage12
  - stage13
  - stage14
bar_job1:
  stage: stage3
bar_seq1_job1:
  stage: stage4
bar_jobx:
  stage: stage5
bar_job2:
  stage: stage12
bar_joby:
  stage: stage13
job1:
  stage: stage6
baz_job1:
  stage: stage7
baz_seq1_job1:
  stage: stage8
baz_seq1_jobx:
  stage: stage5
baz_seq1_joby:
  stage: stage13
baz_job2:
  stage: stage14
"
`;
